std/math/fp2.asm

```diff
@@ -0,0 +1,159 @@
+use std::math::ff::inverse;
+use std::convert::fe;
+use std::convert::int;
+use std::convert::expr;
+use std::field::modulus;
+use std::prover::eval;
+
+// An extension field element, represented as a tuple (a0, a1) of expressions (which evaluate to field elements).
+// An element of F_{p^2} is a polynomial a0 + a1 * x (mod x^2 - 7).
+// The polynomial x^2 - 7 is irreducible in both the Goldilocks and BN254 field
+// (which follows from 7 not having a square root in both fields).
+enum Fp2Expr {
+    Fp2(expr, expr)
+}
+
+// An element of F_{p^2}, but represented as a tuple of *evaluated* field elements.
+enum Fp2Value {
+    Fp2(fe, fe)
+}
+
+let zero_expr = Fp2Expr::Fp2(0, 0);
+let one_expr = Fp2Expr::Fp2(1, 0);
+
+let zero_value = Fp2Value::Fp2(0, 0);
+let one_value = Fp2Value::Fp2(1, 0);
+
+// Extension field addition
+let add_ext: Fp2Expr, Fp2Expr -> Fp2Expr = |a, b| match (a, b) {
+    (Fp2Expr::Fp2(a0, a1), Fp2Expr::Fp2(b0, b1)) => Fp2Expr::Fp2(
+        a0 + b0,
+        a1 + b1
+    )
+};
+
+// Extension field subtraction
+let sub_ext: Fp2Expr, Fp2Expr -> Fp2Expr = |a, b| match (a, b) {
+    (Fp2Expr::Fp2(a0, a1), Fp2Expr::Fp2(b0, b1)) => Fp2Expr::Fp2(
+        a0 - b0,
+        a1 - b1
+    )
+};
+
+// Extension field multiplication
+let mul_ext: Fp2Expr, Fp2Expr -> Fp2Expr = |a, b| match (a, b) {
```

Comment:
Does this actually work correctly if `expr` is e.g. `w^2`?

Reply:
Essentially the question is: Do "create a polynomial ring on top of a feld" and "create an extension field modulo `x^2 - 7`" commute?

Reply:
I'm not sure I understand this question! An extension field element $(a, b) \in \mathbb{F}^2$ just represents $a + bx \in \mathbb{F}_{p^2}$ ($= \mathbb{F}_p[x] \mod x^2 - 7$). The multiplication should work for any pair of field elements (or expression that evaluates to a field element).

Reply:
Yeah, true, when we write $w^2 = w\cdot w$ we actually mean $w(x) * w(x)$ for some $x$.